commit b40dd814b056e86986a4e12b07d81f29cb75443c (from b033783e0f553ef21a20fd2e3e0775b966598646)
Merge: b033783 76bf93e
Author: Scott Lahteine <thinkyhead@users.noreply.github.com>
Date:   Fri Jan 30 18:35:02 2015 -0800

    Merge pull request #1434 from AnHardt/LostInTranslation2
    
    Introduction of a system to feed displays with the correct charsets. Example in language_de.h

diff --git a/Marlin/Configuration.h b/Marlin/Configuration.h
index 3aaa588..f5c218d 100644
--- a/Marlin/Configuration.h
+++ b/Marlin/Configuration.h
@@ -533,6 +533,12 @@ const bool Z_MAX_ENDSTOP_INVERTING = true; // set to true to invert the logic of
 #define ABS_PREHEAT_FAN_SPEED 255   // Insert Value between 0 and 255
 
 //LCD and SD support
+
+// Character based displays can have different extended charsets. If you see strange or missing characters on yours - try an other definition below.
+#define DISPLAY_CHARSET_HD44870_JAPAN  //  "ääööüüß°"
+//#define DISPLAY_CHARSET_HD44870_WESTERN // with better support for European languages. "ÄäÖöÜüß°"
+//#define DISPLAY_CHARSET_DOGM // for full graphics display "ÄäÖÖÜÜß°"
+
 //#define ULTRA_LCD  //general LCD support, also 16x2
 //#define DOGLCD  // Support for SPI LCD 128x64 (Controller ST7565R graphic Display Family)
 //#define SDSUPPORT // Enable SD Card Support in Hardware Console
diff --git a/Marlin/language.h b/Marlin/language.h
index 67fc1b3..defe1e9 100644
--- a/Marlin/language.h
+++ b/Marlin/language.h
@@ -1,6 +1,8 @@
 #ifndef LANGUAGE_H
 #define LANGUAGE_H
 
+#include "Configuration.h"
+
 #define LANGUAGE_CONCAT(M)       #M
 #define GENERATE_LANGUAGE_INCLUDE(M)  LANGUAGE_CONCAT(language_##M.h)
 
@@ -160,6 +162,71 @@
 
 #define MSG_ERR_EEPROM_WRITE                "Error writing to EEPROM!"
 
+// LCD Menu Messages
+
+#define STR_Ae "Ae" // No charset known now. Default will be overruled when we know.
+#define STR_ae "ae"
+#define STR_Oe "Oe"
+#define STR_oe "oe"
+#define STR_Ue "Ue"
+#define STR_ue "ue"
+#define STR_sz "ss"
+#define STR_Deg ""
+#define STR_THERMOMETER "\302"
+
+#ifdef DISPLAY_CHARSET_DOGM
+  #define STR_Ae "\304" // U8glib
+  #define STR_ae "\344"
+  #define STR_Oe "\326"
+  #define STR_oe STR_Oe
+  #define STR_Ue "\334"
+  #define STR_ue STR_Ue
+  #define STR_sz "\337"
+  #define STR_Deg "\260"
+  #define STR_THERMOMETER "\377"
+#endif
+#ifdef DISPLAY_CHARSET_HD44870_JAPAN // HD44870 ROM Code: A00 (Japan)
+  #define STR_ae "\xe1"
+  #define STR_Ae STR_ae
+  #define STR_oe "\357"
+  #define STR_Oe STR_oe
+  #define STR_ue "\365"
+  #define STR_Ue STR_ue
+  #define STR_sz "\342"
+  #define STR_Deg "\271"
+  #define STR_THERMOMETER "\302"
+#endif
+#ifdef DISPLAY_CHARSET_HD44870_WESTERN // HD44870 ROM Code: A02 (Western)
+  #define STR_Ae "\216"
+  #define STR_ae "\204"
+  #define STR_Oe "\211"
+  #define STR_oe "\204"
+  #define STR_Ue "\212"
+  #define STR_ue "\201"
+  #define STR_sz "\160"
+  #define STR_Deg "\337"
+  #define STR_THERMOMETER "\302"
+#endif
+
+/*
+#define TESTSTRING000 "\000\001\002\003\004\005\006\007\010\011\012\013\014\015\016\017"
+#define TESTSTRING020 "\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037"
+#define TESTSTRING040 "\040\041\042\043\044\045\046\047\050\051\052\053\054\055\056\057"
+#define TESTSTRING060 "\060\061\062\063\064\065\066\067\070\071\072\073\074\075\076\077"
+#define TESTSTRING100 "\100\101\102\103\104\105\106\107\110\111\112\113\114\115\116\117"
+#define TESTSTRING120 "\120\121\122\123\124\125\126\127\130\131\132\133\134\135\136\137"
+#define TESTSTRING140 "\140\141\142\143\144\145\146\147\150\151\152\153\154\155\156\157"
+#define TESTSTRING160 "\160\161\162\163\164\165\166\167\170\171\172\173\174\175\176\177"
+#define TESTSTRING200 "\200\201\202\203\204\205\206\207\210\211\212\213\214\215\216\217"
+#define TESTSTRING220 "\220\221\222\223\224\225\226\227\230\231\232\233\234\235\236\237"
+#define TESTSTRING240 "\240\241\242\243\244\245\246\247\250\251\252\253\254\255\256\257"
+#define TESTSTRING260 "\260\261\262\263\264\265\266\267\270\271\272\273\274\275\276\277"
+#define TESTSTRING300 "\300\301\302\303\304\305\306\307\310\311\312\313\314\315\316\317"
+#define TESTSTRING320 "\320\321\322\323\324\325\326\327\330\331\332\333\334\335\336\337"
+#define TESTSTRING340 "\340\341\342\343\344\345\346\347\350\351\352\353\354\355\356\357"
+#define TESTSTRING360 "\360\361\362\363\364\365\366\367\370\371\372\373\374\375\376\377"
+*/
+
 #include LANGUAGE_INCLUDE
 
 #endif //__LANGUAGE_H
diff --git a/Marlin/language_de.h b/Marlin/language_de.h
index 8aa0602..bbad533 100644
--- a/Marlin/language_de.h
+++ b/Marlin/language_de.h
@@ -9,97 +9,96 @@
 #define LANGUAGE_DE_H
 
 #define WELCOME_MSG                         MACHINE_NAME " Bereit."
-#define MSG_SD_INSERTED                     "SDKarte erkannt"
-#define MSG_SD_REMOVED                      "SDKarte entfernt"
-#define MSG_MAIN                            "Hauptmenü"
+#define MSG_SD_INSERTED                     "SDKarte erkannt."
+#define MSG_SD_REMOVED                      "SDKarte entfernt."
+#define MSG_MAIN                            "Hauptmen" STR_ue
 #define MSG_AUTOSTART                       "Autostart"
-#define MSG_DISABLE_STEPPERS                "Stepper abschalt."
-#define MSG_AUTO_HOME                       "Auto Nullpunkt"
-#define MSG_SET_HOME_OFFSETS                "Set home offsets"
-#define MSG_SET_ORIGIN                      "Setze Nullpunkt"
-#define MSG_PREHEAT_PLA                     "Vorwärmen PLA"
-#define MSG_PREHEAT_PLA_N                   "Vorwärmen PLA "
+#define MSG_DISABLE_STEPPERS                "Motoren Aus" // M84
+#define MSG_AUTO_HOME                       "Home" // G28
+#define MSG_SET_HOME_OFFSETS                "Setze Home hier"
+#define MSG_SET_ORIGIN                      "Setze Null hier" //"G92 X0 Y0 Z0" commented out in ultralcd.cpp
+#define MSG_PREHEAT_PLA                     "Vorw" STR_ae "rmen PLA"
+#define MSG_PREHEAT_PLA_N                   "Vorw" STR_ae "rmen PLA "
 #define MSG_PREHEAT_PLA_ALL                 "Vorw. PLA Alle"
 #define MSG_PREHEAT_PLA_BEDONLY             "Vorw. PLA Bett"
-#define MSG_PREHEAT_PLA_SETTINGS            "Vorwärm. PLA Ein."
-#define MSG_PREHEAT_ABS                     "Vorwärmen ABS"
-#define MSG_PREHEAT_ABS_N                   "Vorwärmen ABS "
+#define MSG_PREHEAT_PLA_SETTINGS            "Vorw" STR_ae "rm. PLA Ein."
+#define MSG_PREHEAT_ABS                     "Vorw" STR_ae "rmen ABS"
+#define MSG_PREHEAT_ABS_N                   "Vorw" STR_ae "rmen ABS "
 #define MSG_PREHEAT_ABS_ALL                 "Vorw. ABS Alle"
 #define MSG_PREHEAT_ABS_BEDONLY             "Vorw. ABS Bett"
-#define MSG_PREHEAT_ABS_SETTINGS            "Vorwärm. ABS Ein."
-#define MSG_COOLDOWN                        "Abkühlen"
-#define MSG_SWITCH_PS_ON                    "Switch Power On"
-#define MSG_SWITCH_PS_OFF                   "Switch Power Off"
-#define MSG_EXTRUDE                         "Extrude"
+#define MSG_PREHEAT_ABS_SETTINGS            "Vorw" STR_ae "rm. ABS Ein."
+#define MSG_COOLDOWN                        "Abk" STR_ue "hlen"
+#define MSG_SWITCH_PS_ON                    "Netzteil Ein"
+#define MSG_SWITCH_PS_OFF                   "Netzteil Aus"
 #define MSG_RETRACT                         "Retract"
-#define MSG_MOVE_AXIS                       "Achsen bewegen"
-#define MSG_MOVE_X                          "X bewegen"
-#define MSG_MOVE_Y                          "Y bewegen"
-#define MSG_MOVE_Z                          "Z bewegen"
-#define MSG_MOVE_E                          "Extruder"
-#define MSG_MOVE_01MM                       "0.1mm bewegen"
-#define MSG_MOVE_1MM                        "1mm bewegen"
-#define MSG_MOVE_10MM                       "10mm bewegen"
-#define MSG_SPEED                           "Geschw"
-#define MSG_NOZZLE                          "Düse"
+#define MSG_MOVE_AXIS                       "Bewegen"
+#define MSG_MOVE_X                          "X"
+#define MSG_MOVE_Y                          "Y"
+#define MSG_MOVE_Z                          "Z"
+#define MSG_MOVE_E                          "E"
+#define MSG_MOVE_01MM                       " 0.1 mm"
+#define MSG_MOVE_1MM                        " 1.0 mm"
+#define MSG_MOVE_10MM                       "10.0 mm"
+#define MSG_SPEED                           "Geschw."
+#define MSG_NOZZLE                          "D" STR_ue "se"
 #define MSG_BED                             "Bett"
-#define MSG_FAN_SPEED                       "Lüftergeschw."
+#define MSG_FAN_SPEED                       "L" STR_ue "ftergeschw."
 #define MSG_FLOW                            "Fluss"
 #define MSG_CONTROL                         "Einstellungen"
-#define MSG_MIN                             "\002 Min"
-#define MSG_MAX                             "\002 Max"
-#define MSG_FACTOR                          "\002 Faktor"
+#define MSG_MIN                             STR_THERMOMETER " Min"
+#define MSG_MAX                             STR_THERMOMETER " Max"
+#define MSG_FACTOR                          STR_THERMOMETER " Faktor"
 #define MSG_AUTOTEMP                        "AutoTemp"
 #define MSG_ON                              "Ein"
 #define MSG_OFF                             "Aus"
-#define MSG_PID_P                           "PID-P"
-#define MSG_PID_I                           "PID-I"
-#define MSG_PID_D                           "PID-D"
-#define MSG_PID_C                           "PID-C"
-#define MSG_ACC                             "Acc"
-#define MSG_VXY_JERK                        "Vxy-jerk"
-#define MSG_VZ_JERK                         "Vz-jerk"
-#define MSG_VE_JERK                         "Ve-jerk"
-#define MSG_VMAX                            "Vmax "
+#define MSG_PID_P                           "PID P"
+#define MSG_PID_I                           "PID I"
+#define MSG_PID_D                           "PID D"
+#define MSG_PID_C                           "PID C"
+#define MSG_ACC                             "A"
+#define MSG_VXY_JERK                        "V xy jerk"
+#define MSG_VZ_JERK                         "V z  jerk"
+#define MSG_VE_JERK                         "V e  jerk"
+#define MSG_VMAX                            "V max " // space by purpose
 #define MSG_X                               "x"
 #define MSG_Y                               "y"
 #define MSG_Z                               "z"
 #define MSG_E                               "e"
-#define MSG_VMIN                            "Vmin"
+#define MSG_VMIN                            "V min"
 #define MSG_VTRAV_MIN                       "VTrav min"
-#define MSG_AMAX                            "Amax "
-#define MSG_A_RETRACT                       "A-Retract"
-#define MSG_XSTEPS                          "Xsteps/mm"
-#define MSG_YSTEPS                          "Ysteps/mm"
-#define MSG_ZSTEPS                          "Zsteps/mm"
-#define MSG_ESTEPS                          "Esteps/mm"
+#define MSG_AMAX                            "A max " // space by purpose
+#define MSG_A_RETRACT                       "A Retract"
+#define MSG_XSTEPS                          "X steps/mm"
+#define MSG_YSTEPS                          "Y steps/mm"
+#define MSG_ZSTEPS                          "Z steps/mm"
+#define MSG_ESTEPS                          "E steps/mm"
 #define MSG_TEMPERATURE                     "Temperatur"
 #define MSG_MOTION                          "Bewegung"
 #define MSG_VOLUMETRIC                      "Filament"
-#define MSG_VOLUMETRIC_ENABLED		        "E in mm3"
-#define MSG_FILAMENT_SIZE_EXTRUDER_0        "Fil. Dia. 1"
-#define MSG_FILAMENT_SIZE_EXTRUDER_1        "Fil. Dia. 2"
-#define MSG_FILAMENT_SIZE_EXTRUDER_2        "Fil. Dia. 3"
+#define MSG_VOLUMETRIC_ENABLED		          "E in mm3"
+#define MSG_FILAMENT_SIZE_EXTRUDER_0        "Filament D 1"
+#define MSG_FILAMENT_SIZE_EXTRUDER_1        "Filament D 2"
+#define MSG_FILAMENT_SIZE_EXTRUDER_2        "Filament D 3"
 #define MSG_CONTRAST                        "LCD contrast"
 #define MSG_STORE_EPROM                     "EPROM speichern"
 #define MSG_LOAD_EPROM                      "EPROM laden"
 #define MSG_RESTORE_FAILSAFE                "Standardkonfig."
 #define MSG_REFRESH                         "Aktualisieren"
-#define MSG_WATCH                           "Beobachten"
+#define MSG_WATCH                           "Info"
 #define MSG_PREPARE                         "Vorbereitung"
 #define MSG_TUNE                            "Justierung"
-#define MSG_PAUSE_PRINT                     "Druck anhalten"
-#define MSG_RESUME_PRINT                    "Druck fortsetz"
-#define MSG_STOP_PRINT                      "Druck stoppen"
-#define MSG_CARD_MENU                       "SDKarten Menü"
+#define MSG_PAUSE_PRINT                     "SD-Druck Pause"
+#define MSG_RESUME_PRINT                    "SD-Druck Weiter"
+#define MSG_STOP_PRINT                      "SD-Druck Abbruch"
+#define MSG_CARD_MENU                       "SDKarte"
 #define MSG_NO_CARD                         "Keine SDKarte"
 #define MSG_DWELL                           "Warten..."
-#define MSG_USERWAIT                        "Warte auf Nutzer"
-#define MSG_RESUMING                        "Druck fortsetzung"
-#define MSG_PRINT_ABORTED                   "Print aborted"
-#define MSG_NO_MOVE                         "Kein Zug."
-#define MSG_KILLED                          "KILLED"
-#define MSG_STOPPED                         "GESTOPPT"
+#define MSG_USERWAIT                        "Warte auf Nutzer."
+#define MSG_RESUMING                        "Druck geht weiter"
+#define MSG_PRINT_ABORTED                   "Druck abgebrochen"
+#define MSG_NO_MOVE                         "Motoren Eingesch."
+#define MSG_KILLED                          "KILLED."
+#define MSG_STOPPED                         "ANGEHALTEN."
 #define MSG_CONTROL_RETRACT                 "Retract mm"
 #define MSG_CONTROL_RETRACT_SWAP            "Wechs. Retract mm"
 #define MSG_CONTROL_RETRACTF                "Retract  V"
@@ -109,22 +108,22 @@
 #define MSG_CONTROL_RETRACT_RECOVERF        "UnRet  V"
 #define MSG_AUTORETRACT                     "AutoRetr."
 #define MSG_FILAMENTCHANGE                  "Filament wechseln"
-#define MSG_INIT_SDCARD                     "Init. SD-Card"
-#define MSG_CNG_SDCARD                      "Change SD-Card"
-#define MSG_ZPROBE_OUT                      "Z probe out. bed"
-#define MSG_POSITION_UNKNOWN                "Home X/Y before Z"
+#define MSG_INIT_SDCARD                     "SDKarte erkennen"// Manually initialize the SD-card via user interface
+#define MSG_CNG_SDCARD                      "SDKarte erkennen"// SD-card changed by user. For machines with no autocarddetect. Both send "M21"
+#define MSG_ZPROBE_OUT                      "Sensor ausserhalb"
+#define MSG_POSITION_UNKNOWN                "X/Y vor Z Homen."
 #define MSG_ZPROBE_ZOFFSET                  "Z Offset"
 #define MSG_BABYSTEP_X                      "Babystep X"
 #define MSG_BABYSTEP_Y                      "Babystep Y"
 #define MSG_BABYSTEP_Z                      "Babystep Z"
-#define MSG_ENDSTOP_ABORT                   "Endstop abort"
+#define MSG_ENDSTOP_ABORT                   "Endstop Abbr. Ein"
 
 #ifdef DELTA_CALIBRATION_MENU
-    #define MSG_DELTA_CALIBRATE             "Delta Calibration"
-    #define MSG_DELTA_CALIBRATE_X           "Calibrate X"
-    #define MSG_DELTA_CALIBRATE_Y           "Calibrate Y"
-    #define MSG_DELTA_CALIBRATE_Z           "Calibrate Z"
-    #define MSG_DELTA_CALIBRATE_CENTER      "Calibrate Center"
+    #define MSG_DELTA_CALIBRATE             "Delta Kalibrieren"
+    #define MSG_DELTA_CALIBRATE_X           "Kalibriere X"
+    #define MSG_DELTA_CALIBRATE_Y           "Kalibriere Y"
+    #define MSG_DELTA_CALIBRATE_Z           "Kalibriere Z"
+    #define MSG_DELTA_CALIBRATE_CENTER      "Kalibriere Mitte"
 #endif // DELTA_CALIBRATION_MENU
 
 #endif // LANGUAGE_DE_H

